process/sync/basicForkDetector.go

```diff
@@ -71,55 +82,58 @@ func (bfd *basicForkDetector) AddHeader(header data.HeaderHandler, hash []byte,
 
 	if state == process.BHProcessed {
 		// create a check point and remove all the past headers
-		bfd.mutFork.Lock()
-		bfd.fork.lastCheckpointNonce = bfd.fork.checkpointNonce
-		bfd.fork.lastCheckpointRound = bfd.fork.checkpointRound
-		bfd.fork.checkpointNonce = header.GetNonce()
-		bfd.fork.checkpointRound = int64(header.GetRound())
-		bfd.mutFork.Unlock()
+		if finalHeader == nil || finalHeader.IsInterfaceNil() {
+			bfd.setFinalCheckpoint(bfd.lastCheckpoint())
+		} else if finalHeader.GetNonce() > core.GenesisBlockNonce {
+			bfd.setFinalCheckpoint(&checkpointInfo{nonce: finalHeader.GetNonce(), round: finalHeader.GetRound()})
+			bfd.append(&headerInfo{
+				nonce: finalHeader.GetNonce(),
+				round: finalHeader.GetRound(),
+				hash:  finalHeaderHash,
+				state: process.BHNotarized,
+			})
+		}
+
+		bfd.addCheckpoint(&checkpointInfo{nonce: header.GetNonce(), round: header.GetRound()})
+
 		bfd.removePastHeaders()
 		bfd.removeInvalidHeaders()
+		bfd.removePastCheckpoints()
 	}
 
 	bfd.append(&headerInfo{
 		nonce: header.GetNonce(),
 		round: header.GetRound(),
-		hash:  hash,
+		hash:  headerHash,
 		state: state,
 	})
 
 	probableHighestNonce := bfd.computeProbableHighestNonce()
-
-	bfd.mutFork.RLock()
-	bfd.fork.lastBlockRound = bfd.rounder.Index()
-	bfd.fork.probableHighestNonce = probableHighestNonce
-	bfd.mutFork.RUnlock()
+	bfd.setLastBlockRound(uint64(bfd.rounder.Index()))
+	bfd.setProbableHighestNonce(probableHighestNonce)
 
 	return nil
 }
 
 func (bfd *basicForkDetector) checkBlockValidity(header data.HeaderHandler, state process.BlockHeaderState) error {
-	bfd.mutFork.RLock()
-	roundDif := int64(header.GetRound()) - bfd.fork.lastCheckpointRound
-	nonceDif := int64(header.GetNonce() - bfd.fork.lastCheckpointNonce)
-	bfd.mutFork.RUnlock()
+	roundDif := int64(header.GetRound()) - int64(bfd.finalCheckpoint().round)
+	nonceDif := int64(header.GetNonce()) - int64(bfd.finalCheckpoint().nonce)
+	nextRound := bfd.rounder.Index() + 1
 
 	if roundDif <= 0 {
 		return ErrLowerRoundInBlock
 	}
 	if nonceDif <= 0 {
 		return ErrLowerNonceInBlock
 	}
-	if int64(header.GetRound()) > bfd.rounder.Index() {
+	if int64(header.GetRound()) > nextRound {
```

Comment:
should not be this >= ? as nextRound is Index()+1

Reply:
Firstly it was like this, but I have seen that there are some cases when a quick proposed header, is refused here because of some little out of synchronization in terms of milliseconds between nodes clock. Actually this could happen only for blocks which are comes with flag BHProposed, and they will count only for probable highest nonce. We should check if a fast consensus could also generate this situation also for BHReceived blocks. This could be happen if the proposed header will receive very quick all the signatures needed and the proposer will broadcast this block before some nodes will advance in the next round because of the time clock differences in terms of milliseconds. For now I will keep it like this with a TODO there to analyze if this could generate some attack vectors. Anyway any header received will be checked below for signature and randome seed.